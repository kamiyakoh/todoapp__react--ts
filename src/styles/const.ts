/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';

interface bp {
  tab: number;
  pc: number;
  ex: number;
}
const breakpoints: bp = { tab: 600, pc: 960, ex: 1422 };
export const mq = (bp: keyof bp): string => {
  return `@media (width > ${breakpoints[bp]}px)`;
};

export const pink = css`
  --color: #ff7fbf;
`;
export const blue = css`
  --color: #6fd1ff;
`;
export const yellow = css`
  --color: #ffff6b;
`;
export const green = css`
  --color: #8fff70;
`;
export const size2 = css`
  --size: 2;
`;
export const size3 = css`
  --size: 3;
`;
export const textOrange = css`
  color: #ff6c00;
`;
export const textPink = css`
  color: #ff7fbf;
`;
export const textYellow = css`
  color: #ffff6b;
`;
export const bgLightYellow = css`
  background: #fbfbab;
`;
export const bgLightPink = css`
  background: #ffd9ec;
`;
export const dBlock = css`
  display: block;
`;
export const dInline = css`
  display: inline;
`;
export const dNone = css`
  display: none;
`;
export const fs3 = css`
  font-size: 3rem;
`;
export const fwBold = css`
  font-weight: bold;
`;
export const singleBoard = css`
  width: 100%;
  margin: 2em auto;
  ${mq('pc')} {
    width: 100%;
  }
`;
export const sec = css`
  min-height: calc(100vh - 80px);
  padding: 32px 0;
`;
export const form = css`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  gap: 24px;
  ${mq('tab')} {
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-end;
    gap: 0;
  }

  div {
    * + * {
      margin-top: 8px;
    }
    label + button {
      margin-top: 24px;
    }
  }

  input {
    display: block;
    width: 100%;
  }
`;
export const btn = css`
  cursor: pointer;
  color: var(--color, #fff);
  background-color: transparent;
  border-radius: calc(var(--size, 1) * 2px + 2px);
  border: solid 2px var(--color, #fff);
  font-size: calc(var(--size, 1) * 1rem);
  line-height: 1;
  padding: calc(var(--size, 1) * 2px + 2px);
  border-width: calc(var(--size, 1) * 2px);
`;
export const toastBoard = {
  padding: '1em',
  background: '#202020',
  color: '#fff',
  border: '8px solid #8b4513',
  borderRadius: '3px',
  boxShadow: '0 0 5px #333, 0 0 5px #555 inset',
};
