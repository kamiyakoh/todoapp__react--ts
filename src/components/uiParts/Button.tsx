import type { MouseEventHandler, ReactNode } from 'react';
import type { SerializedStyles } from '@emotion/react';
/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from '@emotion/react';
import { btn } from '../../styles/const';

interface Props {
  isSubmit: boolean;
  btnId?: string;
  cssName?: SerializedStyles;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  children: ReactNode;
}

export const Button = ({ isSubmit, btnId, cssName, onClick, children }: Props): JSX.Element => {
  return (
    <button type={isSubmit ? 'submit' : 'button'} id={btnId} css={[btn, cssName]} onClick={onClick}>
      {children}
    </button>
  );
};
