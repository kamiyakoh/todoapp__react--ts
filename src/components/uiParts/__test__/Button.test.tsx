import type { SerializedStyles } from '@emotion/react';
/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from '@emotion/react';
import renderer from 'react-test-renderer';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import { Button } from '../Button';
import { pink, size2 } from '../../../styles/const';
import { createSerializer, matchers } from '@emotion/jest';

expect.addSnapshotSerializer(createSerializer({ DOMElements: false }));
expect.extend(matchers);

describe('Button Test', () => {
  afterEach(() => {
    cleanup();
  });

  const mockOnClick = jest.fn();

  test.each([
    [true, 'submit'],
    [false, 'button'],
  ])('Props-isSubmit Test', (isSubmit: boolean, expected: string) => {
    render(
      <Button isSubmit={isSubmit} btnId="txt" onClick={mockOnClick}>
        Click me
      </Button>
    );

    const buttonElm = screen.getByRole('button');
    expect(buttonElm).toHaveAttribute('type', expected);
  });

  test.each`
    cssName
    ${undefined}
    ${[pink, size2]}
  `('Props-cssName Test', (cssName: SerializedStyles | SerializedStyles[]) => {
    const tree = renderer
      .create(
        <Button isSubmit={false} btnId="txt" cssName={cssName} onClick={mockOnClick}>
          Click me
        </Button>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('onClick Test', () => {
    render(
      <Button isSubmit={false} btnId="txt" onClick={mockOnClick}>
        Click me
      </Button>
    );
    const buttonElm = screen.getByRole('button');

    fireEvent.click(buttonElm);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });
});
