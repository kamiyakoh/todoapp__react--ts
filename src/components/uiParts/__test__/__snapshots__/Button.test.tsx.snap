// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Test Props-cssName Test 1`] = `
.emotion-0 {
  cursor: pointer;
  color: var(--color, #fff);
  background-color: transparent;
  border-radius: calc(var(--size, 1) * 2px + 2px);
  border: solid 2px var(--color, #fff);
  font-size: calc(var(--size, 1) * 1rem);
  line-height: 1;
  padding: calc(var(--size, 1) * 2px + 2px);
  border-width: calc(var(--size, 1) * 2px);
}

<button
  className="emotion-0"
  id="txt"
  onClick={[MockFunction]}
  type="button"
>
  Click me
</button>
`;

exports[`Button Test Props-cssName Test 2`] = `
.emotion-0 {
  cursor: pointer;
  color: var(--color, #fff);
  background-color: transparent;
  border-radius: calc(var(--size, 1) * 2px + 2px);
  border: solid 2px var(--color, #fff);
  font-size: calc(var(--size, 1) * 1rem);
  line-height: 1;
  padding: calc(var(--size, 1) * 2px + 2px);
  border-width: calc(var(--size, 1) * 2px);
}

.emotion-0 cssName {
  --color: #ff7fbf;
  --size: 2;
}

<button
  className="emotion-0"
  id="txt"
  onClick={[MockFunction]}
  type="button"
>
  Click me
</button>
`;
